File: src/test/java/org/elasticsearch/plugin/readonlyrest/acl/blocks/rules/impl/KibanaAccessRuleTests.java
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */

package org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.impl;

import com.google.common.collect.Sets;
import junit.framework.TestCase;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.RuleExitResult;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.SyncRule;
import org.elasticsearch.plugin.readonlyrest.requestcontext.RequestContext;
import org.elasticsearch.plugin.readonlyrest.acl.domain.KibanaAccess;
import org.elasticsearch.plugin.readonlyrest.settings.rules.KibanaAccessRuleSettings;
import org.elasticsearch.plugin.readonlyrest.utils.esdependent.MockedESContext;
import org.mockito.Mockito;

import java.util.Optional;
import java.util.Set;

import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

/**
 * Created by sscarduzio on 18/01/2017.
 */
public class KibanaAccessRuleTests extends TestCase {

  private RuleExitResult match(Conf configured, Found found, RequestContext rc, boolean involvesIndices) {
    when(rc.involvesIndices()).thenReturn(involvesIndices);
    when(rc.getIndices()).thenReturn(found.indices);
    when(rc.getAction()).thenReturn(found.action);
    when(rc.getUri()).thenReturn(found.uri);
    when(rc.resolveVariable(anyString())).then(invocation -> Optional.of((String) invocation.getArguments()[0]));

    SyncRule r = new KibanaAccessSyncRule(
        new KibanaAccessRuleSettings(configured.accessLevel, configured.kibanaIndex),
        MockedESContext.INSTANCE
    );
    return r.match(rc);
  }

  private RuleExitResult matchRule(KibanaAccess accessLevel, String action) {
    return matchRule(accessLevel, action, Sets.newHashSet(".kibana"));
  }

  private RuleExitResult matchRule(KibanaAccess accessLevel, String action, Set<String> indices) {
    return matchRule(accessLevel, action, indices, ".kibana", false);
  }

  private RuleExitResult matchRule(KibanaAccess accessLevel, String action, Set<String> indices,
                                   String kibanaIndex, boolean involvesIndices) {
    return matchRule(accessLevel, action, indices, kibanaIndex, involvesIndices, "");
  }

  private RuleExitResult matchRule(KibanaAccess accessLevel, String action,
                                  Set<String> indices, String kibanaIndex, boolean involvesIndices, String uri) {
    Conf conf = new Conf();
    conf.accessLevel = accessLevel;
    conf.kibanaIndex = kibanaIndex;

    Found found = new Found();
    found.uri = uri;
    found.action = action;
    found.indices = indices;

    return match(conf, found, Mockito.mock(RequestContext.class), involvesIndices);
  }

  public void testCLUSTER() {
    for (String action : KibanaAccessSyncRule.CLUSTER.getMatchers()) {
      System.out.println("trying " + action + " as RO");
      assertTrue(matchRule(KibanaAccess.RO, action).isMatch());
      System.out.println("trying " + action + " as RW");
      assertTrue(matchRule(KibanaAccess.RW, action).isMatch());
    }
  }

  public void testRO() {
    for (String action : KibanaAccessSyncRule.RO.getMatchers()) {
      System.out.println("trying " + action + " as RO_STRICT");
      assertTrue(matchRule(KibanaAccess.RO_STRICT, action).isMatch());
      System.out.println("trying " + action + " as RO");
      assertTrue(matchRule(KibanaAccess.RO, action).isMatch());
      System.out.println("trying " + action + " as RW");
      assertTrue(matchRule(KibanaAccess.RW, action).isMatch());
    }
  }

  public void testRW() {
    for (String action : KibanaAccessSyncRule.RW.getMatchers()) {
      String a = action.replace("*", "_");

      System.out.println("trying " + a + " as RO_STRICT");
      assertFalse(
          matchRule(KibanaAccess.RO_STRICT, a, Sets.newHashSet(".kibana"), ".kibana", true)
              .isMatch()
      );

      System.out.println("trying " + a + " as RO");
      assertFalse(
          matchRule(KibanaAccess.RO, a, Sets.newHashSet(".kibana"), ".kibana", true)
              .isMatch()
      );

      System.out.println("trying " + a + " as RW");
      assertTrue(
          matchRule(KibanaAccess.RW, a, Sets.newHashSet(".kibana"), ".kibana", true)
              .isMatch()
      );
    }
  }

  public void testROotherIndices() {
    for (String action : KibanaAccessSyncRule.RO.getMatchers()) {
      assertTrue(matchRule(KibanaAccess.RO_STRICT, action, Sets.newHashSet("xxx")).isMatch());
      assertTrue(matchRule(KibanaAccess.RO, action, Sets.newHashSet("xxx")).isMatch());
      assertTrue(matchRule(KibanaAccess.RW, action, Sets.newHashSet("xxx")).isMatch());
    }
  }

  public void testRWotherIndices() {
    for (String action : KibanaAccessSyncRule.RW.getMatchers()) {
      assertFalse(matchRule(KibanaAccess.RO_STRICT, action, Sets.newHashSet("xxx")).isMatch());
      assertFalse(matchRule(KibanaAccess.RO, action, Sets.newHashSet("xxx")).isMatch());
      assertFalse(matchRule(KibanaAccess.RW, action, Sets.newHashSet("xxx")).isMatch());
    }
  }

  public void testROmixedIndices() {
    for (String action : KibanaAccessSyncRule.RO.getMatchers()) {
      assertTrue(matchRule(KibanaAccess.RO_STRICT, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
      assertTrue(matchRule(KibanaAccess.RO, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
      assertTrue(matchRule(KibanaAccess.RW, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
    }
  }

  public void testRWmixedIndices() {
    for (String action : KibanaAccessSyncRule.RW.getMatchers()) {
      assertFalse(matchRule(KibanaAccess.RO_STRICT, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
      assertFalse(matchRule(KibanaAccess.RO, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
      assertFalse(matchRule(KibanaAccess.RW, action, Sets.newHashSet("xxx", ".kibana")).isMatch());
    }
  }

  public void testRWcustomKibanaIndex() {
    String customKibanaIndex = ".kibana-custom";
    Set<String> indices = Sets.newHashSet(customKibanaIndex);

    for (String action : KibanaAccessSyncRule.RW.getMatchers()) {
      System.out.println("trying " + action + " as RO_STRICT");
      assertFalse(matchRule(KibanaAccess.RO_STRICT, action, indices, customKibanaIndex, true).isMatch());

      System.out.println("trying " + action + " as RO");
      assertFalse(matchRule(KibanaAccess.RO, action, indices, customKibanaIndex, true).isMatch());

      System.out.println("trying " + action + " as RW");
      assertTrue(matchRule(KibanaAccess.RW, action, indices, customKibanaIndex, true).isMatch());
    }
  }

  public void testRONonStrictOperations() {
    String customKibanaIndex = ".kibana-custom";
    Set<String> indices = Sets.newHashSet(customKibanaIndex);
    String action = "indices:data/write/index";
    String uri = "/" + customKibanaIndex + "/index-pattern/job";

    System.out.println("trying " + action + " as RO");
    assertFalse(matchRule(KibanaAccess.RO_STRICT, action, indices, customKibanaIndex, true, uri).isMatch());
    assertTrue(matchRule(KibanaAccess.RO, action, indices, customKibanaIndex, true, uri).isMatch());
    assertTrue(matchRule(KibanaAccess.RW, action, indices, customKibanaIndex, true, uri).isMatch());
  }

  class Conf {
    KibanaAccess accessLevel = KibanaAccess.RO;
    String kibanaIndex = ".kibana";
  }

  class Found {
    String action = null;
    Set<String> indices = Sets.newHashSet(".kibana");
    String uri = "";
  }
}

File: corex/src/test/scala/tech/beshu/ror/unit/acl/blocks/rules/KibanaAccessRuleTests.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.unit.acl.blocks.rules


import monix.execution.Scheduler.Implicits.global
import org.scalatest.Matchers._
import org.scalatest.{Inside, WordSpec}
import tech.beshu.ror.utils.TestsUtils.BlockContextAssertion
import tech.beshu.ror.acl.blocks.rules.KibanaAccessRule
import tech.beshu.ror.Constants
import tech.beshu.ror.acl.domain.Header.Name
import tech.beshu.ror.acl.domain.KibanaAccess.{RO, ROStrict, RW}
import tech.beshu.ror.acl.domain._
import tech.beshu.ror.acl.headerValues._
import tech.beshu.ror.mocks.MockRequestContext
import tech.beshu.ror.acl.blocks.rules.Rule.RuleResult.{Fulfilled, Rejected}
import tech.beshu.ror.acl.blocks.{BlockContext, Const, RequestContextInitiatedBlockContext}

import scala.collection.JavaConverters._
import scala.concurrent.duration._
import scala.language.postfixOps

class KibanaAccessRuleTests extends WordSpec with Inside with BlockContextAssertion[KibanaAccessRule.Settings] {

  "A KibanaAccessRule" when {
    "RO action is passed" in {
      Constants.RO_ACTIONS.asScala.map(Action.apply).foreach { action =>
        assertMatchRule(settingsOf(ROStrict), action)()
        assertMatchRule(settingsOf(RO), action)()
        assertMatchRule(settingsOf(RW), action)()
      }
    }
    "CLUSTER action is passed" in {
      Constants.CLUSTER_ACTIONS.asScala.map(Action.apply).foreach { action =>
        assertMatchRule(settingsOf(RO), action)()
        assertMatchRule(settingsOf(RW), action)()
      }
    }
    "RW action is passed" in {
      Constants.RW_ACTIONS.asScala.map(str => Action(str.replace("*", "_")))
        .foreach { action =>
          assertNotMatchRule(settingsOf(ROStrict), action, Set(IndexName(".kibana")))
          assertNotMatchRule(settingsOf(RO), action, Set(IndexName(".kibana")))
          assertMatchRule(settingsOf(RW), action, Set(IndexName(".kibana"))) {
            assertBlockContext(
              responseHeaders = Set(Header(Name.kibanaAccess, RW: KibanaAccess)),
              kibanaIndex = Some(IndexName(".kibana"))
            )
          }
        }
    }
    "RO action is passed with other indices" in {
      Constants.RO_ACTIONS.asScala.map(Action.apply).foreach { action =>
        assertMatchRule(settingsOf(ROStrict), action, Set(IndexName("xxx")))()
        assertMatchRule(settingsOf(RO), action, Set(IndexName("xxx")))()
        assertMatchRule(settingsOf(RW), action, Set(IndexName("xxx")))()
      }
    }
    "RW action is passed with other indices" in {
      Constants.RW_ACTIONS.asScala.map(Action.apply)
        .foreach { action =>
          assertNotMatchRule(settingsOf(ROStrict), action, Set(IndexName("xxx")))
          assertNotMatchRule(settingsOf(RO), action, Set(IndexName("xxx")))
          assertNotMatchRule(settingsOf(RW), action, Set(IndexName("xxx")))
        }
    }
    "RO action is passed with mixed indices" in {
      Constants.RO_ACTIONS.asScala.map(Action.apply).foreach { action =>
        assertMatchRule(settingsOf(ROStrict), action, Set(IndexName("xxx"), IndexName(".kibana")))()
        assertMatchRule(settingsOf(RO), action, Set(IndexName("xxx"), IndexName(".kibana")))()
        assertMatchRule(settingsOf(RW), action, Set(IndexName("xxx"), IndexName(".kibana")))()
      }
    }
    "RW action is passed with mixed indices" in {
      Constants.RW_ACTIONS.asScala.map(Action.apply)
        .foreach { action =>
          assertNotMatchRule(settingsOf(ROStrict), action, Set(IndexName("xxx"), IndexName(".kibana")))
          assertNotMatchRule(settingsOf(RO), action, Set(IndexName("xxx"), IndexName(".kibana")))
          assertNotMatchRule(settingsOf(RW), action, Set(IndexName("xxx"), IndexName(".kibana")))
        }
    }
    "RW action is passed with custom kibana index" in {
      Constants.RW_ACTIONS.asScala.map(Action.apply)
        .foreach { action =>
          assertNotMatchRule(settingsOf(ROStrict, IndexName(".custom_kibana")), action, Set(IndexName(".custom_kibana")))
          assertNotMatchRule(settingsOf(RO, IndexName(".custom_kibana")), action, Set(IndexName(".custom_kibana")))
          assertMatchRule(settingsOf(RW, IndexName(".custom_kibana")), action, Set(IndexName(".custom_kibana"))) {
            assertBlockContext(
              responseHeaders = Set(Header(Name.kibanaAccess, RW: KibanaAccess)),
              kibanaIndex = Some(IndexName(".custom_kibana"))
            )
          }
        }
    }
    "non strict operations (1)" in {
      testNonStrictOperations(
        customKibanaIndex = IndexName(".custom_kibana"),
        action = Action("indices:data/write/index"),
        uriPath = UriPath("/.custom_kibana/index-pattern/job")
      )
    }
    "non strict operations (2)" in {
      testNonStrictOperations(
        customKibanaIndex = IndexName(".custom_kibana"),
        action = Action("indices:data/write/delete"),
        uriPath = UriPath("/.custom_kibana/index-pattern/nilb-auh-filebeat-*")
      )
    }
    "non strict operations (3)" in {
      testNonStrictOperations(
        customKibanaIndex = IndexName(".custom_kibana"),
        action = Action("indices:admin/template/put"),
        uriPath = UriPath("/_template/kibana_index_template%3A.kibana")
      )
    }
    "non strict operations (4)" in {
      testNonStrictOperations(
        customKibanaIndex = IndexName(".custom_kibana"),
        action = Action("indices:data/write/update"),
        uriPath = UriPath("/.custom_kibana/doc/index-pattern%3A895e56e0-d873-11e8-bd16-3dcc5288c87b/_update?")
      )
    }
  }

  private def testNonStrictOperations(customKibanaIndex: IndexName, action: Action, uriPath: UriPath): Unit = {
    assertNotMatchRule(settingsOf(ROStrict, customKibanaIndex), action, Set(customKibanaIndex), Some(uriPath))
    assertMatchRule(settingsOf(RO, customKibanaIndex), action, Set(customKibanaIndex), Some(uriPath)){
      assertBlockContext(
        responseHeaders = Set(Header(Name.kibanaAccess, RO: KibanaAccess)),
        kibanaIndex = Some(customKibanaIndex)
      )
    }
    assertMatchRule(settingsOf(RW, customKibanaIndex), action, Set(customKibanaIndex), Some(uriPath)) {
      assertBlockContext(
        responseHeaders = Set(Header(Name.kibanaAccess, RW: KibanaAccess)),
        kibanaIndex = Some(customKibanaIndex)
      )
    }
  }

  private def assertMatchRule(settings: KibanaAccessRule.Settings, action: Action, indices: Set[IndexName] = Set.empty, uriPath: Option[UriPath] = None)
                             (blockContextAssertion: BlockContext => Unit = defaultOutputBlockContextAssertion(settings)) =
    assertRule(settings, action, indices, uriPath, Some(blockContextAssertion))

  private def assertNotMatchRule(settings: KibanaAccessRule.Settings, action: Action, indices: Set[IndexName] = Set.empty, uriPath: Option[UriPath] = None) =
    assertRule(settings, action, indices, uriPath, blockContextAssertion = None)

  private def assertRule(settings: KibanaAccessRule.Settings,
                         action: Action,
                         indices: Set[IndexName],
                         uriPath: Option[UriPath] = None,
                         blockContextAssertion: Option[BlockContext => Unit]) = {
    val rule = new KibanaAccessRule(settings)
    val requestContext = MockRequestContext(
      action = action,
      indices = indices,
      uriPath = uriPath.getOrElse(UriPath(""))
    )
    val blockContext = RequestContextInitiatedBlockContext.fromRequestContext(requestContext)
    val result = rule.check(requestContext, blockContext).runSyncUnsafe(1 second)
    blockContextAssertion match {
      case Some(assertOutputBlockContext) =>
        inside(result) { case Fulfilled(outBlockContext) =>
          assertOutputBlockContext(outBlockContext)
        }
      case None =>
        result should be(Rejected)
    }
  }

  private def settingsOf(access: KibanaAccess, kibanaIndex: IndexName = IndexName(".kibana")) = {
    KibanaAccessRule.Settings(access, Const(kibanaIndex), kibanaMetadataEnabled = true)
  }

  private def defaultOutputBlockContextAssertion(settings: KibanaAccessRule.Settings): BlockContext => Unit =
    (blockContext: BlockContext) => {
      assertBlockContext(responseHeaders = Set(Header(Name.kibanaAccess, settings.access)))(blockContext)
    }

}
